New component `esp_wifi_remote`

david-cermak opened the pull request:
## Initial version

### TODO

** This version **

- [x] Add README and docs
- [x] Cleanup git history
- [x] new component stuff (changelog, versions)

** Next versions **

* Add APIs from `esp_wifi_he.h` 
* Add cross-check that slave and host wifi config matches
* Add coexistence of the wifi_remote and wifi_native configurations

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/espressif/esp-protocols?pullRequest=516) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/espressif/esp-protocols?pullRequest=516) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: david-cermak<br/>:x: mantriyogesh<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/espressif/esp-protocols?pullRequest=516) it.</sub>

mantriyogesh commented on components/esp_wifi_remote/wifi_remote_net.c:
```diff
@@ -0,0 +1,90 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <esp_private/wifi.h>
```
private may create issues?

david-cermak replied:
We need to keep it as is, since we're using wifi netifs as esp_hosted channels. we need to use `wifi_internal_` APIs for it.

mantriyogesh commented:
Hello @david-cermak ,

I had reviewed the changes, looks fine. But let me test these and confirm.
Some small code changes I had to do while running, which I will attach as patch here..

espressif-abhikroy commented on components/esp_wifi_remote/wifi_remote_rpc.c:
```diff
@@ -0,0 +1,232 @@
+
+esp_err_t remote_esp_wifi_set_config(wifi_interface_t interface, wifi_config_t *conf)
+{
+#if 0
+    uint8_t *param = (uint8_t *)conf;
+    uint32_t checksum = 0; // TODO: generate a random number and add it to both
+    for (int i = 0; i < sizeof(wifi_config_t); ++i) {
+        checksum += param[i];
+    }
+
+    // transmit the sensitive parameters over a secure channel
+//    s_params_tx(s_params_channel, param, sizeof(wifi_config_t));
+
+    // add only a checksum to the RPC
+    return esp_hosted_wifi_set_config(interface, checksum);
+#endif
```
Is this a test code?

david-cermak replied:
I'll remove it.

david-cermak replied:
removed

euripedesrocha commented on components/esp_wifi_remote/test/smoke_test/sdkconfig.ci.6:
```diff
@@ -0,0 +1 @@
+CONFIG_SLAVE_IDF_TARGET_ESP32C6=y
```
Have you considered some alternative to define target under test from CI matrix?

david-cermak replied:
I don't think we could actually use github matrix.
the only optiohn is to use ${ENV ...} and set it up from environmnet, but the current method is cleaner IMO

david-cermak replied:
Left as is, but renamed configs from numbers to `slave_esp32xx`

euripedesrocha commented on components/esp_wifi_remote/wifi_remote_net.c:
```diff
@@ -0,0 +1,90 @@
+}
+
+
+esp_err_t esp_wifi_internal_set_sta_ip(void)
```
Should we add a LOG stating that this is not supported?

david-cermak replied:
I'll have to check where it is used. but will definitely add an error/debug log

david-cermak replied:
This is still a shortcut at the moment, this API is called from netif layers and notifies wifi library that we got an IP.
I've added a `TODO:...` comment, that we need to transfer this info to the slave device and all `esp_wifi_internal_set_sta_ip()` there

alisitsyn commented on components/esp_wifi_remote/scripts/check_headers.sh:
```diff
@@ -0,0 +1,22 @@
+#!/usr/bin/env bash
+set -e
+
+check_files="
```
minor: is this better to define the list of headers for check differently? from separate list file or wifi_remote__build.yml file?

david-cermak replied:
Make sense! will add a single source of true

david-cermak replied:
I've refactored the script and checker, so It's not needed anymore to keep a list of files to check. I'm doing it in the python script, which only populates the list of files to check when generating them.

alisitsyn reviewed (commented):
Left just minor comments.

kostaond commented on components/esp_wifi_remote/scripts/README.md:
```diff
@@ -0,0 +1,38 @@
+
+Based on extracted WiFi function prototypes we generate
+
+* `esp_wifi_remote_api.h` -- declares all wifi APIs with `esp_wifi_remote...()` prefix
```
I think there should be comment in all generated files that "they were autogenerated".

david-cermak replied:
will add a comment that the files are autogenerated

david-cermak replied:
All auto-generated files should have this comment now

espressif-abhikroy commented on components/esp_wifi_remote/README.md:
```diff
@@ -0,0 +1,20 @@
+
+## Dependencies on `esp_wifi`
+
+Public API needs to correspond exactly to the `esp_wifi` API. Some of the internal types depend on the actual wifi target, as well as some default configuration values. Therefore it's easier to maintain consistency between this component and the exact version of `esp_wifi` automatically in CI:
+
+We extract function prototypes from `esp_wifi.h` and use them to generate `esp_wifi_remote` function declarations (and forwarding the definitions to the underlying RPC component -- `esp_hosted`)
+We process the local `esp_wifi_types_native.h` and replace `CONFIG_IDF_TARGET` to `CONFIG_SLAVE_IDF_TARGET` and `CONFIG_SOC_WIFI_...` to `CONFIG_SLAVE_...`.
+Similarly we process `esp_wifi`'s Kconfig, so the dependencies are on the slave target and slave SOC capabilities.
```
A lot of things are explained in these lines. It might be easier to read and understand if they are turned into bullet points.

david-cermak replied:
I'll rewrite it to be more structured text

espressif-abhikroy commented on ci/ignore_astyle.txt:
```diff
@@ -0,0 +1 @@
+components/esp_wifi_remote/include/esp_wifi_types_native.h
```
Why is astyle ignored on this header file?

david-cermak replied:
good point, I'll add a comment and explain why it's in the ignore list

david-cermak replied:
added comment

kostaond commented on components/esp_wifi_remote/scripts/copyright_header.h:
```diff
@@ -0,0 +1,5 @@
+/*
+ * SPDX-FileCopyrightText: 2024 Espressif Systems (Shanghai) CO LTD
```
How it is handled when year changes?

david-cermak replied:
I kept the workflow, but added a description in the checker what to do and which file to update if it fails.

kostaond commented on components/esp_wifi_remote/README.md:
```diff
@@ -0,0 +1,20 @@
+
+[![Component Registry](https://components.espressif.com/components/espressif/esp_wifi_remote/badge.svg)](https://components.espressif.com/components/espressif/esp_wifi_remote)
+
+This component wraps the public API of `esp_wifi` and provides a set of the same function calls namespaces with `esp_wifi_remote` prefix that translate to RPC calls to another target device (called `slave` device) which executes the appropriate `esp_wifi` APIs.
```
I also think the README is not very user oriented. For example, I'm not sure how to use this component. If I wanted to use it, do I need to `#include "esp_wifi_remote.h"`?

david-cermak replied:
I will add a description on:
* why we have it
* how to use it
* high level step-by-step guidelines (also the HW requirements)

david-cermak replied:
README updated and added one user oriented paragraph to the beginning

gabsuren commented on .github/workflows/wifi_remote__build.yml:
```diff
@@ -0,0 +1,48 @@
+        run: |
+          ${IDF_PATH}/install.sh --enable-pytest
+          . ${IDF_PATH}/export.sh
+          python ./ci/build_apps.py ./components/esp_wifi_remote/${{matrix.test.path}} -vv --preserve-all
```
What about target tests?

david-cermak replied:
no target tests for this version.

david-cermak replied:
I'll add target tests in upcoming versions.

david-cermak commented on components/esp_wifi_remote/include/esp_wifi_remote.h:
```diff
@@ -0,0 +1,31 @@
+#include "esp_wifi_remote_api.h"
+#endif
+
+// TODO: Move this to private include
```
I haven't move it to private header, I'll move it.

david-cermak replied:
Moved and created [esp_wifi_remote_private.h](https://github.com/espressif/esp-protocols/blob/60c9d62da6d7145040714d8f9d806c4d26bc5718/components/esp_wifi_remote/esp_wifi_remote_private.h)

david-cermak commented on components/esp_wifi_remote/idf_component.yml:
```diff
@@ -0,0 +1,8 @@
+dependencies:
+  idf:
+    version: '5.3'
+  espressif/esp_hosted:
```
I'll remove `esp_hosted` dependency for the initial version, but will add it later with more strict deps

david-cermak commented on components/esp_wifi_remote/idf_component.yml:
```diff
@@ -0,0 +1,8 @@
+description: Utility wrapper for esp_wifi functionality on remote targets
+dependencies:
+  idf:
+    version: '5.3'
```
related wifi changes are only supported in 5.3

david-cermak commented on components/esp_wifi_remote/scripts/parse_header.py:
```diff
@@ -0,0 +1,301 @@
+            f.write(f'    return {remote_func_name}({names});\n')
+            f.write('}\n')
+
+    with open('esp_hosted_mock.c', 'w') as f:
+        with open('esp_hosted_mock.h', 'w') as h:
```
will do

david-cermak commented on components/esp_wifi_remote/wifi_remote_net.c:
```diff
@@ -0,0 +1,90 @@
+
+void esp_wifi_internal_free_rx_buffer(void *buffer)
+{
+    if (buffer) {
+        free(buffer);
+    }
```
will do

david-cermak commented on components/esp_wifi_remote/README.md:
```diff
@@ -0,0 +1,20 @@
+
+Public API needs to correspond exactly to the `esp_wifi` API. Some of the internal types depend on the actual wifi target, as well as some default configuration values. Therefore it's easier to maintain consistency between this component and the exact version of `esp_wifi` automatically in CI:
+
+We extract function prototypes from `esp_wifi.h` and use them to generate `esp_wifi_remote` function declarations (and forwarding the definitions to the underlying RPC component -- `esp_hosted`)
```
will read the readme and rewrite it.

david-cermak commented on components/esp_wifi_remote/test/smoke_test/components/esp_hosted/include/esp_hosted_mock.h:
```diff
@@ -0,0 +1,82 @@
+esp_err_t esp_hosted_wifi_connectionless_module_set_wake_interval(uint16_t wake_interval);
+esp_err_t esp_hosted_wifi_force_wakeup_acquire(void);
+esp_err_t esp_hosted_wifi_force_wakeup_release(void);
+esp_err_t esp_hosted_wifi_set_country_code(const char *country, _Bool ieee80211d_enabled);
```
I'll check if there's an easy way to tell my scripts to use `bool` instead, but `_Bool` is the actual underlying type

david-cermak replied:
I'll check where it's defined

david-cermak replied:
it's actually as simple as a macro defined in `<sdtbool.h>`:
```cpp
#define bool	_Bool
#define true	1
#define false	0
```

david-cermak commented on components/esp_wifi_remote/scripts/parse_header.py:
```diff
@@ -0,0 +1,301 @@
+                    if line.strip().startswith('config SOC_WIFI_'):
+                        if 'config SOC_WIFI_SUPPORTED' in line:
+                            test_config += 1    # if WiFi supported for this target, test it as a slave
+                            open(f'sdkconfig.ci.{test_config}', 'w').write(f'CONFIG_SLAVE_IDF_TARGET_{slave_target.upper()}=y\n')
```
make sense! will call them based on the slave target. `sdkconfig.ci.slave_esp32c3`

david-cermak commented on components/esp_wifi_remote/scripts/README.md:
```diff
@@ -0,0 +1,38 @@
+
+* `esp_wifi_remote_api.h` -- declares all wifi APIs with `esp_wifi_remote...()` prefix
+* `esp_wifi_with_remote.c` -- defines all original wifi APIs that forward calls for `esp_wifi_remote` -- This file is used only for targets with no WiFi capabilities, so the original WiFi APIs could be called the same way as on WiFi targets. (this file is not compiled for targets with WiFi caps)
+* `esp_wifi_remote_with_hosted.h` -- defines `static inline` functions that directly forward calls from `esp_wifi_remote` to `esp_hosted`
```
good point, I'll add a link to GH repo

mantriyogesh commented on components/esp_wifi_remote/wifi_remote_rpc.c:
```diff
@@ -0,0 +1,232 @@
+
+esp_err_t remote_esp_wifi_scan_get_ap_records(uint16_t *number, wifi_ap_record_t *ap_records)
+{
+    return esp_hosted_wifi_scan_get_ap_records(number, ap_records);
```
We can give __weak__ definitions with return value failed, may add a debug level  log.
esp_hosted calls can be removed then

david-cermak replied:
okay agree

david-cermak replied:
`WEAK` definitions of `esp_wifi_remote` API are generated now

david-cermak commented on components/esp_wifi_remote/wifi_remote_net.c:
```diff
@@ -0,0 +1,90 @@
+esp_err_t esp_wifi_remote_channel_rx(void *h, void *buffer, void *buff_to_free, size_t len)
+{
+    assert(h);
+    if (h == s_channel[0]) {
+        assert(s_rx_fn[0]);
+        return s_rx_fn[0](buffer, len, buff_to_free);
+    }
+    if (h == s_channel[1]) {
+        assert(s_rx_fn[1]);
+        return s_rx_fn[1](buffer, len, buff_to_free);
+    }
```
will apply.

david-cermak commented:
Hi guys, I've addressed most of the comments, could you please take another look and check mainly the unresolved discussions if it's okay with you? I likely  won't be able to resolve it *correctly* with the initial version, though.

mantriyogesh commented:
[hosted_hooked_with_esp_wifi_remote.patch](https://github.com/espressif/esp-protocols/files/14769399/hosted_hooked_with_esp_wifi_remote.patch)

david-cermak commented:
> [hosted_hooked_with_esp_wifi_remote.patch](https://github.com/espressif/esp-protocols/files/14769399/hosted_hooked_with_esp_wifi_remote.patch)

Thanks for sharing the patch. I've updated the PR accordingly and simplified the component a little:
* removed `esp_wifi_remote_slave_init()` altogether while making the channel operations public -- helps `esp_hosted` to implement it based on public channels
* defined most of the channel operations `WEAK`, so `esp_hosted` can update them
* made the minor modification and updates (logs, nullptr checks)

david-cermak commented:
Thanks @euripedesrocha for the review. I'm taking the liberty of merging this now, I'll address the outstanding comments in the upcoming version. Thanks everyone for having a look!
